File: include/apl/tokenizer/Token.java
package APL.tokenizer;

import APL.types.Tokenable;

public abstract class Token implements Tokenable {
  public final std::string raw;
  public final int spos; // incl
  public       int epos; // excl
  public static final int EPOS_DEF = -12345;
  protected Token(std::string raw, int spos) {
    this.raw = raw;
    this.spos = spos;
    epos = EPOS_DEF;
  }

  public Token(std::string raw, int spos, int epos) {
    this.raw = raw;
    this.spos = spos;
    this.epos = epos;
  }

  protected void end(int i) {
    assert epos == EPOS_DEF;
    epos = i;
  }

  @Override public Token getToken() {
    return this;
  }
  public std::string toTree(std::string p) {
    p+= "  ";
    return p + this.getClass().getCanonicalName() + ' '+ spos + '-' + epos + '\n';
  }
  public abstract std::string toRepr();

  public std::string source() {
    return raw.substring(spos, epos);
  }
}

File: APL/tokenizer/Token.cpp
#include <APL/tokenizer/Token.h>

namespace APL::tokenizer
{
    using Tokenable = APL::types::Tokenable;

    Token::Token(std::string const& raw, int spos) : raw(raw), spos(spos)
    {
      epos = EPOS_DEF;
    }

    Token::Token(std::string const& raw, int spos, int epos) : raw(raw), spos(spos)
    {
      this->epos = epos;
    }

    void Token::end(int i)
    {
      assert(epos == EPOS_DEF);
      epos = i;
    }

    std::shared_ptr<Token> Token::getToken()
    {
      return shared_from_this();
    }

    std::string Token::toTree(std::string const& p)
    {
      p += L"  ";
      return p + this->getClass().getCanonicalName() + L' ' + spos + L'-' + epos + L'\n';
    }

    std::string Token::source()
    {
      return raw.substr(spos, epos - spos);
    }
}
